marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/mapping.sjs

```diff
@@ -0,0 +1,78 @@
+/**
+ Copyright 2012-2019 MarkLogic Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+'use strict';
+
+const DataHubSingleton = require('/data-hub/5/datahub-singleton.sjs');
+
+// define constants for caching expensive operations
+const dataHub = DataHubSingleton.instance();
+
+const collections = ['http://marklogic.com/data-hub/mapping-artifact'];
+const databases = [dataHub.config.STAGINGDATABASE, dataHub.config.FINALDATABASE];
+const permissions = [xdmp.permission(dataHub.consts.DATA_HUB_MAPPING_WRITE_ROLE, 'update'), xdmp.permission(dataHub.consts.DATA_HUB_MAPPING_READ_ROLE, 'read')];
+const requiredProperties = ['name', 'targetEntity', 'selectedSource'];
+
+function getNameProperty() {
+    return 'name';
+}
+
+function getEntityNameProperty() {
+    return 'targetEntity';
+}
+
+function getSelectedSourceProperty() {
+    return 'selectedSource';
+}
+
+function getVersionProperty() {
+    return null;
+}
+
+function getCollections() {
+    return collections;
+}
+
+function getStorageDatabases() {
+    return databases;
+}
+
+function getPermissions() {
+    return permissions;
+}
+
+function getArtifactNode(artifactName, artifactVersion) {
+    // Currently there is no versioning for loadData artifacts
+    const results = cts.search(cts.andQuery([cts.collectionQuery(collections[0]), cts.jsonPropertyValueQuery('name', artifactName)]));
+    return fn.head(results);
+}
+
+function validateArtifact(artifact) {
+    const missingProperties = requiredProperties.filter((propName) => !artifact[propName]);
+    if (missingProperties.length) {
+        return new Error(`Missing the following required properties: ${JSON.stringify(missingProperties)}`);
+    }
+    return artifact;
+}
+
+module.exports = {
```

Comment:
@hao1st can we please add a test for validateArtifact?

Reply:
i will talk to @ryanjdew tomorrow, and we can simply the validateArtifact implemented in core.sjs. Actually we do not need to have this one in the artifact module.

Reply:
@ryanjdew In the core.sjs, 
```
function validateArtifact(artifactType, artifactName, artifact) {
    const artifactLibrary =  getArtifactTypeLibrary(artifactType);
    const validatedArtifact = artifactLibrary.validateArtifact(artifact, artifactName);
    if (validatedArtifact instanceof Error) {
        returnErrToClient(400, 'BAD REQUEST', validatedArtifact.message);
    }
    return validatedArtifact;
}

I think we can move the function used in loadData.sjs to core.sjs

```
function validateArtifact(artifact) {
    const missingProperties = requiredProperties.filter((propName) => !artifact[propName]);
    if (missingProperties.length) {
        return new Error(`Missing the following required properties: ${JSON.stringify(missingProperties)}`);
    }
    return artifact;
}
```
Different artifacts just need to expose "requiredProperties" so that we do not need to override/duplicate validateArtifact with the same implmentation.

Reply:
@hao1st  We'll want to extend the validation to be able to encompass more than the required properties. 

e.g., for mapping, we'll want to enable the validation to eventually do the validation XPath expressions, etc. We already have the libraries for that from what we used in QS. We'll want to eventually call those from the mapping artifact library.

Reply:
@bsrikan OK, I added the validation test case for mapping test.